node/overseer/overseer-gen/proc-macro/src/impl_builder.rs

```diff
@@ -578,3 +594,51 @@ pub(crate) fn impl_task_kind(info: &OverseerInfo) -> proc_macro2::TokenStream {
 
 	ts
 }
+
+/// This is a helper function that produces placeholder types for all
+/// subsystems and baggage fields of the overseer
+fn create_all_placeholder_idents(
+	subsystems_names: &[Ident],
+	baggage_names: &[Ident],
+) -> Vec<Ident> {
+	let mut result = Vec::<Ident>::with_capacity(subsystems_names.len() + baggage_names.len());
+
+	result.extend(
+		subsystems_names
+			.iter()
+			.map(|field_name| format_ident!("__{}", field_name)),
+	);
+	result.extend(
+		baggage_names
+			.iter()
+			.map(|field_name| format_ident!("__{}", field_name)),
+	);
+	result
+}
+
+/// This helper function is used to create a replacement for a placeholder types
+/// replacing a field matches specific index with a replacement ident
+/// For example, we have `<A, B, C>` and we want to produce `<A, Init<T>, C>`,
+/// so we call this function for `idx=1` and replacement ident as `Init<T>`
+fn produce_setter_generic_permutation<T>(
+	field_type: T,
+	field_idx: usize,
+	placeholder_types: &[Ident],
```

Comment:
I don't quite get why we need both `field_type` and `field_idx`? Couldn't you just drop `field_type`, avoid all the preprocessing of `placeholder_types` (skipping the `field_idx` actually) and do smth along the lines of `builder_generic_ty[idx] = format_ident!("NEW");`?